src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs

```diff
@@ -325,6 +325,28 @@ internal virtual NamedTypeSymbol LookupMetadataType(ref MetadataTypeName emitted
                 }
             }
 
+            if (isTopLevel && MetadataHelpers.DecodeFileType(emittedTypeName.UnmangledTypeName) is (not -1 and var ordinal, var typeName))
+            {
+                // also do a lookup for file types from source.
+                namespaceOrTypeMembers = scope.GetTypeMembers(typeName);
+                foreach (var named in namespaceOrTypeMembers)
+                {
+                    if (named is SourceMemberContainerTypeSymbol { AssociatedSyntaxTree: SyntaxTree tree }
+                        && named.DeclaringCompilation.GetSyntaxTreeOrdinal(tree) == ordinal
+                        && (emittedTypeName.ForcedArity == -1 || emittedTypeName.ForcedArity == emittedTypeName.InferredArity)
+                        && emittedTypeName.InferredArity == named.Arity)
```

Comment:
We decided offline that it's probably intentional to not call the overload which takes arity, since the currently used overload results in fewer allocations.